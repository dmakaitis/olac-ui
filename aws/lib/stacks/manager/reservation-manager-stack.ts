import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import {Construct} from 'constructs';
import {Function} from 'aws-cdk-lib/aws-lambda';
import {Duration} from "aws-cdk-lib";

export interface ReservationManagerStackProps extends cdk.StackProps {

    saveReservationFunction: Function;

    payPal: {
        apiBase: string,
        clientId: string,
        secret: string
    };

}

export class ReservationManagerStack extends cdk.Stack {

    readonly apiGetNewReservationIdFunction: Function;
    readonly apiSaveReservationFunction: Function;
    readonly apiSaveReservationAdminFunction: Function;

    constructor(scope: Construct, id: string, props: ReservationManagerStackProps) {
        super(scope, id, props);

        this.apiGetNewReservationIdFunction = new lambda.Function(this, 'GetNewReservationId', {
            description: 'Generate a new reservation ID',
            runtime: lambda.Runtime.NODEJS_22_X,
            code: lambda.Code.fromAsset('./dist/lambda/manager/reservation/new-reservation-id'),
            handler: 'new-reservation-id.handler',
        });

        this.apiSaveReservationFunction = new Function(this, 'SaveReservationApi', {
            description: 'Saves a new reservation to the datastore (for public use)',
            runtime: lambda.Runtime.NODEJS_22_X,
            code: lambda.Code.fromAsset('./dist/lambda/manager/reservation/save-reservation'),
            handler: 'save-reservation.apiHandler',
            environment: {
                SAVE_RESERVATION_FUNCTION: props.saveReservationFunction.functionName,
                PAYPAL_API_BASE: props.payPal.apiBase,
                PAYPAL_CLIENT_ID: props.payPal.clientId,
                PAYPAL_SECRET: props.payPal.secret
            },
            timeout: Duration.seconds(10)
        });
        props.saveReservationFunction.grantInvoke(this.apiSaveReservationFunction);

        this.apiSaveReservationAdminFunction = new Function(this, 'SaveReservationAdminApi', {
            description: 'Saves a new reservation to the datastore (for event coordinator use)',
            runtime: lambda.Runtime.NODEJS_22_X,
            code: lambda.Code.fromAsset('./dist/lambda/manager/reservation/save-reservation-admin'),
            handler: 'save-reservation-admin.apiHandler',
            environment: {
                SAVE_RESERVATION_FUNCTION: props.saveReservationFunction.functionName,
                PAYPAL_API_BASE: props.payPal.apiBase,
                PAYPAL_CLIENT_ID: props.payPal.clientId,
                PAYPAL_SECRET: props.payPal.secret
            }
        });
        props.saveReservationFunction.grantInvoke(this.apiSaveReservationAdminFunction);
    }
}