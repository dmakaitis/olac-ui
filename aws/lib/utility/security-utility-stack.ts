import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as cognito from 'aws-cdk-lib/aws-cognito';
import {Construct} from 'constructs';

export interface SecurityUtilityStackProps extends cdk.StackProps {
    environment: string,
    userPoolId: string,
    clientId: string
}

export class SecurityUtilityStack extends cdk.Stack {

    readonly whoAmIFunction: lambda.Function;

    readonly adminUserGroup: cognito.CfnUserPoolGroup;
    readonly eventCoordinatorUserGroup: cognito.CfnUserPoolGroup;

    constructor(scope: Construct, id: string, props: SecurityUtilityStackProps) {
        super(scope, id, props);

        this.adminUserGroup = new cognito.CfnUserPoolGroup(this, 'EventUserGroup', {
            userPoolId: props.userPoolId,

            // the properties below are optional
            description: `OLAC Admin Group - ${props.environment}`,
            groupName: `olac-admin-${props.environment.toLowerCase()}`,
        });
        this.eventCoordinatorUserGroup = new cognito.CfnUserPoolGroup(this, 'EventCoordinatorUserGroup', {
            userPoolId: props.userPoolId,

            // the properties below are optional
            description: `OLAC Event Coordinator Group - ${props.environment}`,
            groupName: `olac-event-coordinator-${props.environment.toLowerCase()}`,
        });

        this.whoAmIFunction = new lambda.Function(this, 'WhoAmI', {
            description: 'Returns the identity of the caller',
            runtime: lambda.Runtime.NODEJS_16_X,
            code: lambda.Code.fromAsset('./lambda/utility/security'),
            handler: 'whoami.handler',
            environment: {
                USER_POOL_ID: props.userPoolId,
                CLIENT_ID: props.clientId,
                ADMIN_GROUP_NAME: this.adminUserGroup.groupName || '__UNDEFINED__',
                EVENT_COORDINATOR_GROUP_NAME: this.eventCoordinatorUserGroup.groupName || '__UNDEFINED__'
            }
        });
    }
}
