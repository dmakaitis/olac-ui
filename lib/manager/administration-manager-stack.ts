import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import {Construct} from 'constructs';

export interface AdministrationManagerStackProps extends cdk.StackProps {
    getConfigFunction: lambda.Function
}

export class AdministrationManagerStack extends cdk.Stack {

    readonly getClientConfigFunction: lambda.Function;

    constructor(scope: Construct, id: string, props: AdministrationManagerStackProps) {
        super(scope, id, props);

        this.getClientConfigFunction = new lambda.Function(this, 'GetClientConfig', {
            description: 'Retrieve the client configuration for the UI',
            runtime: lambda.Runtime.NODEJS_16_X,
            code: lambda.Code.fromAsset('./lambda/manager/administration'),
            handler: 'get-client-config.handler',
            environment: {
                GET_CONFIG_FUNCTION: props.getConfigFunction.functionName
            },
        });

        props.getConfigFunction.grantInvoke(this.getClientConfigFunction);
    }
}