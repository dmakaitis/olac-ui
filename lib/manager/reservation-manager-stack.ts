import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import {Construct} from 'constructs';
import {Function} from 'aws-cdk-lib/aws-lambda';

export interface ReservationManagerStackProps extends cdk.StackProps {

    saveReservationFunction: Function

}

export class ReservationManagerStack extends cdk.Stack {

    readonly apiGetNewReservationIdFunction: Function;
    readonly apiSaveReservationFunction: Function;
    readonly apiSaveReservationAdminFunction: Function;

    constructor(scope: Construct, id: string, props: ReservationManagerStackProps) {
        super(scope, id, props);

        this.apiGetNewReservationIdFunction = new lambda.Function(this, 'GetNewReservationId', {
            description: 'Generate a new reservation ID',
            runtime: lambda.Runtime.NODEJS_16_X,
            code: lambda.Code.fromAsset('./lambda/manager/reservation'),
            handler: 'new-reservation-id.handler',
        });

        this.apiSaveReservationFunction = new Function(this, 'SaveReservationApi', {
            description: 'Saves a new reservation to the datastore (for public use)',
            runtime: lambda.Runtime.NODEJS_16_X,
            code: lambda.Code.fromAsset('./lambda/manager/reservation'),
            handler: 'save-reservation.apiHandler',
            environment: {
                SAVE_RESERVATION_FUNCTION: props.saveReservationFunction.functionName
            }
        });
        props.saveReservationFunction.grantInvoke(this.apiSaveReservationFunction);

        this.apiSaveReservationAdminFunction = new Function(this, 'SaveReservationAdminApi', {
            description: 'Saves a new reservation to the datastore (for event coordinator use)',
            runtime: lambda.Runtime.NODEJS_16_X,
            code: lambda.Code.fromAsset('./lambda/manager/reservation'),
            handler: 'save-reservation-admin.apiHandler',
            environment: {
                SAVE_RESERVATION_FUNCTION: props.saveReservationFunction.functionName
            }
        });
        props.saveReservationFunction.grantInvoke(this.apiSaveReservationAdminFunction);
    }
}