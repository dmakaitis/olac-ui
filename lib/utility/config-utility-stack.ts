import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import {Construct} from 'constructs';

export interface ConfigUtilityStackProps extends cdk.StackProps {
    payPal: {
        apiBase: string,
        clientId: string,
        donationButtonId: string
    },
    cognito: {
        domain: string,
        clientId: string,
        redirectUri: string
    }
}

export class ConfigUtilityStack extends cdk.Stack {

    readonly getConfigLambda: lambda.Function;

    constructor(scope: Construct, id: string, props: ConfigUtilityStackProps) {
        super(scope, id, props);

        this.getConfigLambda = new lambda.Function(this, 'GetConfiguration', {
            description: 'Returns the application configuration',
            runtime: lambda.Runtime.NODEJS_18_X,
            code: lambda.Code.fromAsset('./lambda/utility/config'),
            handler: 'get.handler',
            environment: {
                PAYPAL_API_BASE: props.payPal.apiBase,
                PAYPAL_CLIENT_ID: props.payPal.clientId,
                PAYPAL_DONATION_BUTTON_ID: props.payPal.donationButtonId,
                COGNITO_DOMAIN: props.cognito.domain,
                COGNITO_CLIENT_ID: props.cognito.clientId,
                COGNITO_REDIRECT_URI: props.cognito.redirectUri
            },
        });
    }
}
