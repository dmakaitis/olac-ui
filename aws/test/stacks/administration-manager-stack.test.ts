import {Template} from "aws-cdk-lib/assertions";
import {App} from "aws-cdk-lib";
import {AdministrationManagerStack} from "../../lib/stacks/manager/administration-manager-stack";
import {Function} from "aws-cdk-lib/aws-lambda"
import {ConfigUtilityStack} from "../../lib/stacks/utility/config-utility-stack";

let app: App;
let configFunction: Function;

describe("AdministrationManagerStack", () => {
    beforeEach(() => {
        app = new App();

        const configStack = new ConfigUtilityStack(app, "ConfigUtilityStack", {
            cognito: {
                clientId: "test-cognito-client-id",
                domain: "test.domain.com",
                redirectUri: "https://test.domain.com/redirect"
            },
            payPal: {
                apiBase: "https://test.paypal.com/api",
                clientId: "test-paypal-client-id",
                donationButtonId: "test-paypal-button-id"
            }
        });
        configFunction = configStack.getConfigLambda
    })

    test("Administration manager stack creation", () => {
        const stack = new AdministrationManagerStack(app, "AdministrationManagerStack", {
            getConfigFunction: configFunction
        });
        const template = Template.fromStack(stack);
        expect(template.toJSON()).toMatchSnapshot()///
    })
})
