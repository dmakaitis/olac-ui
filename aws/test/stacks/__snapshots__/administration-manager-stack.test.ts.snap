// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdministrationManagerStack Administration manager stack creation 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "GetClientConfigF0ACB15B": {
      "DependsOn": [
        "GetClientConfigServiceRoleDefaultPolicy3AB1E58B",
        "GetClientConfigServiceRole9C378730",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "a6ec9890667375016476cc9ae10e8e0946950eca43b0398e51242e8041a94c8e.zip",
        },
        "Description": "Retrieve the client configuration for the UI",
        "Environment": {
          "Variables": {
            "GET_CONFIG_FUNCTION": {
              "Fn::ImportValue": "ConfigUtilityStack:ExportsOutputRefGetConfigurationB56ECE29714C453E",
            },
          },
        },
        "Handler": "get-client-config.handler",
        "Role": {
          "Fn::GetAtt": [
            "GetClientConfigServiceRole9C378730",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "GetClientConfigServiceRole9C378730": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GetClientConfigServiceRoleDefaultPolicy3AB1E58B": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::ImportValue": "ConfigUtilityStack:ExportsOutputFnGetAttGetConfigurationB56ECE29Arn1B5E8961",
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::ImportValue": "ConfigUtilityStack:ExportsOutputFnGetAttGetConfigurationB56ECE29Arn1B5E8961",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GetClientConfigServiceRoleDefaultPolicy3AB1E58B",
        "Roles": [
          {
            "Ref": "GetClientConfigServiceRole9C378730",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
